pkg/controller/cronjob/cronjob_controllerv2.go

```diff
@@ -0,0 +1,640 @@
+	})
+
+	cronJobsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+		UpdateFunc: jm.updateCronJob,
+		DeleteFunc: func(obj interface{}) {
+			jm.enqueueController(obj)
+		},
+	})
+
+	return jm, nil
+}
+
+// Run starts the main goroutine responsible for watching and syncing jobs.
+func (jm *ControllerV2) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer jm.queue.ShutDown()
+
+	klog.Infof("Starting cronjob controller v2")
+	defer klog.Infof("Shutting down cronjob controller v2")
+
+	go jm.enqueueAllExistingCronJobs()
+
+	if !cache.WaitForNamedCacheSync("cronjob", stopCh, jm.jobListerSynced, jm.cronJobListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(jm.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// enqueueAllExistingCronJobs enqueues all the cronjobs and is thread-safe.
+func (jm *ControllerV2) enqueueAllExistingCronJobs() {
+	cronjobs, err := jm.cronJobLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to list cronjobs at the start of cronjob controller, error: %v", err))
+		return
+	}
+	for _, cronjob := range cronjobs {
+		jm.enqueueController(cronjob)
+	}
+	return
+}
+
+func (jm *ControllerV2) worker() {
+	for jm.processNextWorkItem() {
+	}
+}
+
+func (jm *ControllerV2) processNextWorkItem() bool {
+	key, quit := jm.queue.Get()
+	if quit {
+		return false
+	}
+	defer jm.queue.Done(key)
+	requeueAfter, err := jm.sync(key.(string))
+	switch {
+	case err != nil:
+		utilruntime.HandleError(fmt.Errorf("error syncing CronJobController %v, requeuing: %v", key.(string), err))
+		jm.queue.Add(key)
+	case requeueAfter != nil:
+		jm.queue.AddAfter(key, *requeueAfter)
+	default:
+		jm.queue.Done(key)
+	}
+	return true
+}
+
+func (jm *ControllerV2) sync(cronJobKey string) (*time.Duration, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(cronJobKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cronJob, err := jm.cronJobLister.CronJobs(ns).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// may be cronjob is deleted, dont need to requeue this key
+		klog.V(2).InfoS("cronjob not found, may be it is deleted", "cronjob", klog.KRef(ns, name), "err", err)
+		return nil, nil
+	case err != nil:
+		// for other transient apiserver error requeue with exponential backoff
+		return nil, err
+	}
+
+	var jobSelector labels.Selector
+	if len(cronJob.Spec.JobTemplate.Labels) == 0 {
+		jobSelector = labels.Everything()
+	} else {
+		jobSelector = labels.Set(cronJob.Spec.JobTemplate.Labels).AsSelector()
+	}
+	jobList, err := jm.jobLister.List(jobSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	jobsToBeReconciled := []batchv1.Job{}
+
+	for _, job := range jobList {
+		// If it has a ControllerRef, that's all that matters.
+		if controllerRef := metav1.GetControllerOf(job); controllerRef != nil && controllerRef.Name == name {
```

Comment:
jobLister is listing from all namespaces, right?
So we also should compare namespace.

There should also be a test catching that.

Reply:
controllerRef does not have a Namespace key since owner references are within a namespace. But I factored the code into a function, used namespaced lister, and added unit tests for comparing the namespace, PTAL, thanks!